rfcs/20200712-tfrt-kernel-fallback.md

```diff
@@ -0,0 +1,647 @@
+*   Proposed implementation order for these methods:
+    [here](https://github.com/annarev/community/blob/tfrt_kernel_fallback_rfc/rfcs/20200712-tfrt-kernel-fallback/kernel_to_unsupported.md)
+
+Based on these estimates, we can support >= 423 kernels. Note that this number
+is just based on the `OpKernelConstruction`/`OpKernelContext` coverage that we
+can provide. It doesn't take into consideration other issues we might face.
+
+### TFRT Integration Goals
+
+We want to support executing a [BEF](https://github.com/tensorflow/runtime/blob/master/documents/binary_executable_format.md ) file
+on mobile device that calls kernels using Kernel Fallback mechanism. Users will
+be able to generate a BEF file based on a saved model and we will provide a
+script to create it.
+
+We might also want to support running ops using TFRT eager mode (that is, add a
+custom
+[OpHandler](https://github.com/tensorflow/runtime/blob/3c7a1ea02c87325f1b47aebb24b3ca6e84e7e7e7/include/tfrt/core_runtime/op_handler.h#L47)).
+
+## Non-goals
+
+*   Supporting all existing ops. `OpKernelContext` surface is quite large and
+    implementing all of it would require a significant amount of time. Instead,
+    we will start by adding most common and easy functionality. If certain
+    functionality is only used by a handful of kernels, it might make sense to
+    implement TFRT native kernels or rely on runtime fallback instead. One notable example is
+    [ResourceMgr](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/framework/resource_mgr.h;l=152?q=ResourceMgr).
+    We might support it later, but it is definitely not first priority due to
+    extra effort required.
+*   Gradients would not be supported by the first iteration of Kernel Fallback,
+    but we might revisit it later.
+*   Exact details of TFRT integration are still being worked out by TFRT and TF
+    Lite teams. Since these teams might change the plan, exact details are not a
+    part of this doc. The take away is that we will integrate kernel fallback
+    following the approach they decide on.
+
+## Motivation
+
+Currently, [TF Lite](https://www.tensorflow.org/lite) supports a
+[limited set of ops](https://www.tensorflow.org/lite/guide/ops_compatibility).
+As the range and variety of applications grows, it becomes essential to grow the
+pool of available ops as well, ideally supporting everything that fully-fledged
+TensorFlow supports now.
+
+However, supporting TensorFlow ops on mobile devices presents some challenges.
+Specifically, binary size on mobile platforms should be restricted. TF Lite team
+provided us with the following *ideal* numbers:
+
+*   100-200k overhead to call TF kernels
+*   20k / kernel marginal size
+
+To get closer to the size restrictions we plan to define a call path from TFRT
+to TensorFlow kernels that minimizes the amount of generated code.
+
+## User Benefit
+
+Running more kernels on mobile devices would allow TensorFlow Lite users to implement
```

Comment:
I'd also add that a reduced binary size will be beneficial to users that currently use TensorFlow Lite's experimental TensorFlow Select ops (https://www.tensorflow.org/lite/guide/ops_select), or users that do not use the experimental feature because of that reason.

Reply:
Updated.